lib/wallet/src/Cardano/Wallet/DB/Store/Wallets/Store.hs

```diff
@@ -21,66 +19,60 @@ module Cardano.Wallet.DB.Store.Wallets.Store
 import Cardano.Wallet.DB.Store.Transactions.Model
     ( DeltaTxSet (..), mkTxSet )
 import Cardano.Wallet.DB.Store.Wallets.Model
     ( DeltaTxWalletsHistory (..) )
 import Control.Applicative
     ( liftA2 )
-import Control.Exception
-    ( SomeException (..) )
-import Control.Monad.Class.MonadThrow
-    ( MonadThrow, throwIO )
 import Data.DBVar
     ( Store (..) )
-import Data.Delta
-    ( Base, Delta )
+import Database.Persist.Sql
+    ( SqlPersistT )
 
 import qualified Cardano.Wallet.DB.Store.Meta.Model as TxMetaStore
 
+
 mkStoreTxWalletsHistory
-    :: (Monad m, MonadThrow m)
-    => Store m DeltaTxSet
-    -> Store m DeltaTxMetaHistory
-    -> Store m DeltaTxWalletsHistory
+    :: Store (SqlPersistT IO) DeltaTxSet
+    -> QueryStore (SqlPersistT IO) QueryTxMeta DeltaTxMetaHistory
+    -> Store (SqlPersistT IO) DeltaTxWalletsHistory
 mkStoreTxWalletsHistory storeTransactions storeMeta =
-    let load = liftA2 (,)
-            <$> loadS storeTransactions
-            <*> loadS storeMeta
-        write = \(txSet,txMetaHistory) -> do
+    let load =
+            liftA2 (,)
+                <$> loadS storeTransactions
+                <*> loadS (store storeMeta)
+        write = \(txSet, txMetaHistory) -> do
             writeS storeTransactions txSet
-            writeS storeMeta txMetaHistory
+            writeS (store storeMeta) txMetaHistory
         update ma delta =
-            let (mTxSet,mWmetas) = (fst <$> ma, snd <$> ma)
+            let (mTxSet, mWmetas) = (fst <$> ma, snd <$> ma)
             in  case delta of
-            RollbackTxWalletsHistory slot -> do
-                wmetas <- loadWhenNothing mWmetas storeMeta
-                updateS storeMeta (Just wmetas)
-                    $ TxMetaStore.Rollback slot
-                let ( _metas', toBeDeletedTxSet)
-                        = TxMetaStore.rollbackTxMetaHistory slot wmetas
-                updateS storeTransactions mTxSet
-                    $ DeleteTxs toBeDeletedTxSet
-
-            ExpandTxWalletsHistory wid cs -> do
-                wmetas <- loadWhenNothing mWmetas storeMeta
-                updateS storeTransactions mTxSet
-                    $ Append
-                    $ mkTxSet
-                    $ fst <$> cs
-                updateS storeMeta (Just wmetas)
-                    $ TxMetaStore.Expand
-                    $ mkTxMetaHistory wid cs
-    in Store { loadS = load, writeS = write, updateS = update }
-
--- | Call 'loadS' from a 'Store' if the value is not already in memory.
-loadWhenNothing
-    :: (Monad m, MonadThrow m, Delta da)
-    => Maybe (Base da) -> Store m da -> m (Base da)
-loadWhenNothing (Just a) _ = pure a
-loadWhenNothing Nothing store =
-    loadS store >>= \case
-        Left (SomeException e) -> throwIO e
-        Right a -> pure a
+                    RollbackTxWalletsHistory slot -> do
+                        tbd <- case mWmetas of
+                            Nothing -> getMetasAfterSlot storeMeta slot
```

Comment:
Wait — there seems to be an off-by-one error here. 😳 

The `TxMetaStore.*` roll back all transaction `meta`s which satisfy `txMetaSlot meta > slot`. For example, here the condition in `rollbackTxMetaHistory`:

https://github.com/input-output-hk/cardano-wallet/blob/c79149f35a5b5c3c6a2c03e7b3713380ea2935d2/lib/wallet/src/Cardano/Wallet/DB/Store/Meta/Model.hs#L96-L99

However, despite its name, `getMetasAfterSlot` returns all `meta` with the condition `txMetaSlot meta >= slot`. (See comment there.) 😳 


Darn, and it looks like `prop_StoreUpdates` does not catch this mistake, because it always supplies `Just` as the argument to `updateS`. 🤔 One more reason to revisit this argument, as you already pointed out.

https://github.com/input-output-hk/cardano-wallet/blob/897f40acd4c33f0bc77fb7ad942ff6c8a2da7714/lib/dbvar/src/Test/DBVar.hs#L78

Reply:
thanks! this was really bad